docs/serialization.md

```diff
@@ -0,0 +1,89 @@
+# Serialization
+
+Contracts use JSON format on the transport layer.
+
+All three kinds of types (`TypeRef`) have defined serialization formats. Regardless of kind, a `TypeRef` with `nullable = true` is serialized to a JSON `null`.
+
+## `TypeRef::Generic`
+
+Never serialized. Final structures need to have resolved generics to a statically known type.
+
+## `TypeRef::Internal`
+
+These are always composite objects and thus are serialized to a JSON object. Key being the property name and value being the serialized value.
+
+## `TypeRef::Known`
+
+### `KnownType::Object`
+
+TODO: how
+
+### `KnownType::String`
+
+Serialized to a JSON string.
+
+### `KnownType::Guid`
+
+Serialized to a JSON string with the [UUID v4](https://datatracker.ietf.org/doc/html/rfc4122) format.
+
+### `KnownType::Uri`
+
+TODO: what standard?
+
+### `KnownType::Boolean`
+
+Serialized to a JSON boolean.
+
+### `KnownType::UInt8`, `KnownType::Int8`, `KnownType::Int16`, `KnownType::UInt16`, `KnownType::Int32`, `KnownType::UInt32`, `KnownType::Int64`, `KnownType::UInt64`
+
+Serialized to a JSON number with respected signedness, integer bit size, and no decimal part.
+
+### `KnownType::Float32`, `KnownType::Float64`
+
+Serialized to a JSON number with respected floating point bit size.
+
+### `KnownType::DateOnly`
+
+Serialized to a JSON string with the date format `yyyy-MM-dd`.
+
+### `KnownType::TimeOnly`
+
+Serialized to a JSON string with the time format `hh:mm:ss.SSS` (1ms precision).
+
+### `KnownType::DateTimeOffset`
+
+Serialized to a JSON string with the format `yyyy-MM-dd'T'hh:mm:ss.SSS±ZZ:ZZ` or `yyyy-MM-dd'T'hh:mm:ss.SSS'Z'`.
+
+### `KnownType::TimeSpan`
+
+Serialized to a JSON string with the format `dd.hh:mm:ss.SSS` with an optional leading `-` (1ms precision). TODO: how many days?
+
+### `KnownType::Array`
+
+Serialized to a JSON array where values are consequent serialized elements.
+
+### `KnownType::Map`
+
+Serialized to a JSON object TODO: what if keys are not strings?
+
+### `KnownType::Query`, `KnownType::Command`, `KnownType::Operation`
+
+Not serializable. It is assumed that CQRS methods do not end up in payloads.
+
+### `KnownType::CommandResult`
+
+Serialized to a JSON object with the following fields:
+
+- `WasSuccessful` with a JSON boolean as value
```

Comment:
Btw, is this field redundant? Isn't this an invariant with respect to `ValidationErrors` $(|\text{ValidationErrors}| = 0) \equiv \text{WasSuccessful}$?

Reply:
It is redundant and we're thinking about removing it (but when will it happen - I can't tell).